### PR TITLE
Coprarna

### PR DESCRIPTION
**Pull request by eggzilla**

### CODE REVIEW

**Comment by eggzilla** [comment]
Still WIP please dont merge yet

File: tools/rna_tools/coprarna/coprarna.xml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    <param argument="--nooi" truevalue="--nooi" falsevalue="" checked="False" type="boolean" label="No focus on the organism of interest" help="If set then the CopraRNA2 prediction mode is set not to focus on the organism of interest (def:off)"/>
+    <param argument="--enrich" truevalue="--enrich" falsevalue="" checked="False" type="boolean" label="DAVID-WS functional enrichment" help="If set DAVID-WS functional enrichment is calculated with given number of top predictions (def:off)"/>
+    <param argument="--enrich" type="integer" value="0" label="DAVID-WS functional enrichment" help="If set DAVID-WS functional enrichment is calculated with given number of top predictions (def:off)" optional="true"/>
+    <param argument="--root" type="float" value="1" label="Root function" help="Specifies root function to apply to the weights (def:1)"/>
+  </inputs>
+  <outputs>
+    <data format="csv" name="result" from_work_dir="./CopraRNA_result.csv" label="CopraRNA result csv"/>
+    <data format="txt" name="options" from_work_dir="./CopraRNA_option_file.csv" label="CopraRNA option log"/>
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: The log file should be optional or just put onto stdout. An extra element in the Galaxy is probably not needed for the normal biologist.

File: tools/rna_tools/coprarna/coprarna.xml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    <data format="pdf" name="conservationheatmap" from_work_dir="./sRNA_conservation_heatmap.pdf" label="sRNA conservation heatmap"/>
+    <data format="pdf" name="functionheatmap" from_work_dir="./Enrichment/enriched_heatmap_big.pdf" label="Functional enrichment heatmap"/>
+    <data format="tsv" name="clusters" from_work_dir="./cluster.tab" label="clusters"/>
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: this is called "tabular" if it referes to a tab separated file

File: tools/rna_tools/coprarna/coprarna.xml [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    <param argument="--cop1" truevalue="--cop1" falsevalue="" checked="False" type="boolean" label="CopraRNA1 prediction" help="Switch for CopraRNA1 prediction (def:off)"/>
+    <param argument="--nooi" truevalue="--nooi" falsevalue="" checked="False" type="boolean" label="No focus on the organism of interest" help="If set then the CopraRNA2 prediction mode is set not to focus on the organism of interest (def:off)"/>
+    <param argument="--enrich" truevalue="--enrich" falsevalue="" checked="False" type="boolean" label="DAVID-WS functional enrichment" help="If set DAVID-WS functional enrichment is calculated with given number of top predictions (def:off)"/>
+    <param argument="--enrich" type="integer" value="0" label="DAVID-WS functional enrichment" help="If set DAVID-WS functional enrichment is calculated with given number of top predictions (def:off)" optional="true"/>
+    <param argument="--root" type="float" value="1" label="Root function" help="Specifies root function to apply to the weights (def:1)"/>
+  </inputs>
+  <outputs>
+    <data format="csv" name="result" from_work_dir="./CopraRNA_result.csv" label="CopraRNA result csv"/>
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: What is the difference between csv and tsv in Coprana? Is coprana really mixing tab and comma separated files.

File: tools/rna_tools/coprarna/coprarna.xml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      <param name="--cores" value="1"/>
+      <param name="--winsize" value="150"/>
+      <param name="--maxbpdist" value="100"/>
+      <param name="--enrich" value="200"/>
+      <param name="--cons" value="0"/>
+      <output name="result" file="CopraRNA_result.csv"/>
+      <output name="options" file="CopraRNA_option_file.csv"/>
+      <output name="conservationheatmap" file="sRNA_conservation_heatmap.pdf"/>
+      <output name="clusters" file="cluster.tab"/>
+      <output name="functionheatmap" file="enriched_heatmap_big.pdf"/>
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: you can not "diff" a pdf file, so you might want to use another comparison method, like sim_size

File: tools/rna_tools/coprarna/coprarna.xml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+<![CDATA[
+             
+**What it does**
+
+Phylogenetic target prediction for prokaryotic trans-acting small RNAs
+
+CopraRNA is a tool for sRNA target prediction. It computes whole genome target predictions by combination of distinct whole genome IntaRNA predictions. As input CopraRNA requires at least 3 homologous sRNA sequences from 3 distinct organisms in FASTA format. Furthermore, each organisms' genome has to be part of the NCBI Reference Sequence (RefSeq) database (i.e. it should have exactly this NZ_* or this NC_XXXXXX format where * stands for any character and X stands for a digit between 0 and 9). Depending on sequence length (target and sRNA), amount of input organisms and genome sizes, CopraRNA can take up to 24h or longer to compute. In most cases it is significantly faster. It is suggested to run CopraRNA on a machine with at least 8 GB of memory.
+
+**Options**
+-help : help
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: This help is not really useful, as it is already explained in the prams.

File: tools/rna_tools/coprarna/coprarna.xml [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+
+**Output**
+List of conserved identified interactions
+
+    ]]>
+
+  </help>
+  <citations>
+    <citation type="doi">10.1073/pnas.1303248110</citation>
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: 2?

**Comment by bgruening** [comment]
.shed.yml file is missing to trigger the tests.

**Comment by eggzilla** [comment]
Thx, as always for reviewing!!

File: tools/rna_tools/coprarna/test-data/CopraRNA_option_file.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+sRNA FASTA:copra.fa
+nt upstream:75
```
Comment (bgruening) [inline tools/rna_tools/coprarna/test-data/CopraRNA_option_file.txt]: This file is not used, right?

**Comment by eggzilla** [comment]
Hey, i tested also with planemo serve, I think it should be ready to merge.

File: tools/rna_tools/coprarna/coprarna.xml [hunk 8]
```diff
@@ -5,14 +5,18 @@
   <requirements>
     <requirement type="package" version="2.1.1">coprarna</requirement>
   </requirements>
+  <stdio>
+    <exit_code range="1:" level="fatal" description="Failing exit code."/>
+    <exit_code range=":-1" level="fatal" description="Failing exit code."/>
+  </stdio>
   <command detect_errors="aggressive"><![CDATA[
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: if you use detect_errors you can remove the stdio tag, at least in this case how you use it

File: tools/rna_tools/coprarna/coprarna.xml [hunk 9]
```diff
@@ -52,12 +54,13 @@
...
     <data format="pdf" name="functionheatmap" from_work_dir="./Enrichment/enriched_heatmap_big.pdf" label="Functional enrichment heatmap"/>
-    <data format="txt" name="warnings" from_work_dir="warnings" label="Warnings"/>
   </outputs>
   <tests>
-    <test>
+    <test expect_exit_code="1" expect_failure="true">
       <param name="srnaseq" value="copra.fa"/>
-      <output name="warnings" file="warnings"/>
+      <assert_stderr>
+        <has_text text="Error: The input file" />
```
Comment (bgruening) [inline tools/rna_tools/coprarna/coprarna.xml]: Oh, I see ... mh...

**Comment by bgruening** [comment]
Cool thanks!

**Pull request merged**